HW2/P4/driver.py

```diff
@@ -15,15 +15,59 @@
 from timer import Timer
 import threading
 
+def median_3x3_worker(A, B, offset, step, iterations, sems) : 
+    for i in range(iterations):
+        # Acquire appropriate n-1 and n+1 semaphores
+        # This will only be available if iteration i-1 is done for threads
+        # n-1 and n+1
+
+        # Acquiring semaphore in this way will allow us to begin an iteration of filtering
+        # asynchronously
+
+        # Note that this logic is robust enough for N=2 and N=1, though
+        # it will do some unrequired semaphore operations.
+        lower_sem = sems[(offset - 1) % step][1]
+        upper_sem = sems[(offset + 1) % step][0]
+
+        lower_sem.acquire()
+        upper_sem.acquire()
+       
+        # Perform filtering starting at line n and for every Nth line after
+        filtering.median_3x3(A, B, offset, step)
+
+        # swap direction of filtering
+        A, B = B, A
+
+        # Indicate to threads n-1 and n+1 that iteration i is done.
+        (self_sem_1, self_sem_2) = sems[offset]
+        self_sem_1.release()
+        self_sem_2.release()
+
```

Comment:
Interesting... I used (num_threads * iteration) number of locks to make coherent,
but your code seems to be using num_threads number of semaphores..
I wonder which strategy is more efficient :)